Added ase Nudged Elastic Band in the NewtonNet recipe

kumaranu opened the pull request:
## Summary of Changes

\>>
Target here is to add ase's NEB as the double ended transition state search method for NewtonNet recipe.

- I need to make sure that geodesic code is added as a preprocessor for the double ended transition state search.
- NEB will take the output of Geodesic
- Sella will them run from the top point of NEB.
- Tests will need to be added for all the files keeping the coverage of more than 99% before taking it for review.
- Currently, there is a lot of unstructured code in the ts.py and in the test. This will need to be cleaned up to match with the rest of the recipe's in QuAcc.
- The summary dictionaries will need to be written so that the data can be passed back to the database.
- Also, the requirements will be added in the NewtonNet's recipe, however, it is a general problem where ideally the calculator can be changed. If a decision is made later on to move it out of NewtonNet then things will (hopefully) be still usable a general path optimization process and can be added along with the static, relax, freq, and thermo jobs. But, right now, NewtonNet seems like an OK place for this to be put in.

 <<

### Checklist

- [ ] I have read the ["Guidelines" section](https://quantum-accelerators.github.io/quacc/dev/contributing.html#guidelines) of the contributing guide. Don't lie! 😉
- [ ] My PR is on a custom branch and is _not_ named `main`.
- [ ] I have added relevant, comprehensive [unit tests](https://quantum-accelerators.github.io/quacc/dev/contributing.html#unit-tests).

### Notes

- Your PR will likely not be merged without proper and thorough tests.
- If you are an external contributor, you will see a comment from [@buildbot-princeton](https://github.com/buildbot-princeton). This is solely for the maintainers.
- When your code is ready for review, ping one of the [active maintainers](https://quantum-accelerators.github.io/quacc/about/contributors.html#active-maintainers).

buildbot-princeton commented:
Can one of the admins verify this patch?

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+
+def geodesic_interpolate_wrapper(
+    r_p_atoms: Atoms,
+    nimages: int = 17,
```
Probably should pick a more sensible default, e.g. 15 or 20?

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+def geodesic_interpolate_wrapper(
+    r_p_atoms: Atoms,
+    nimages: int = 17,
+    sweep: bool | None = None,
```
Seems like this should just be `bool = False`

kumaranu replied:
Hi @Andrew-S-Rosen ,
This seems like an interesting use of default value structures. It decides whether to call sweep or smooth depending on the size of the system but only when the value provided is None. Otherwise choose whatever `bool` is provided as the argument.
Please let me know if it makes sense. I have followed a structure that was used in the geodesic_interpolate's repository but they just didn't have a wrapper function like this.
Thanks.

Andrew-S-Rosen replied:
In that case, it would likely be clearer to have something like `sweep: bool | Literal["auto"] = "auto"`.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+    microiter: int = 20,
+    scaling: float = 1.7,
+    friction: float = 1e-2,
+    dist_cutoff: float = 3,
```
The default value should be 3.0 instead of 3.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+    # Read the initial geometries.
+    symbols = r_p_atoms[0].get_chemical_symbols()
+
+    X = [conf.get_positions() for conf in r_p_atoms]
```
More descriptive name needed.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+
+    # First redistribute number of images. Perform interpolation if too few and subsampling if too many images are given
+    raw = redistribute(symbols, X, nimages, tol=tol * 5)
+    if save_raw is not None:
```
This should presumably be `if save_raw:` with the aforementioned `bool = False` change.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+    # First redistribute number of images. Perform interpolation if too few and subsampling if too many images are given
+    raw = redistribute(symbols, X, nimages, tol=tol * 5)
+    if save_raw is not None:
+        write_xyz(save_raw, symbols, raw)
```
In quacc, we never write out files to the current working directory because that is not necessarily the same as the directory where the calculations are run. We will need to revisit this to ensure it is written to the proper directory.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+    if sweep is None:
+        sweep = len(symbols) > 35
+    try:
+        if sweep:
+            smoother.sweep(tol=tol, max_iter=maxiter, micro_iter=microiter)
+        else:
+            smoother.smooth(tol=tol, max_iter=maxiter)
```
This should instead be:

```python
smoother.sweep(tol=tol, max_iter=maxiter, micro_iter=micro_iter if sweep else None)
```

(or whatever the default is if it's not `None`).

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+    smoother = Geodesic(symbols, raw, scaling, threshold=dist_cutoff, friction=friction)
+    if sweep is None:
+        sweep = len(symbols) > 35
+    try:
+        if sweep:
+            smoother.sweep(tol=tol, max_iter=maxiter, micro_iter=microiter)
+        else:
+            smoother.smooth(tol=tol, max_iter=maxiter)
+    finally:
+        # Save the smoothed path to output file. try block is to ensure output is saved if one ^C the process, or there is an error
```
Generally, one should avoid the use of `finally` statements. Use a proper `try`/`except` block instead.

Also, typo: if one ^C the process

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+            smoother.smooth(tol=tol, max_iter=maxiter)
+    finally:
+        # Save the smoothed path to output file. try block is to ensure output is saved if one ^C the process, or there is an error
+        write_xyz(output, symbols, smoother.path)
```
Same comment as above about writing out to the current working directory, which is not necessarily the same as the results/runtime directory.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+    return symbols, smoother.path
+
+
+def setup_images(logdir: str, xyz_r_p: str, n_intermediate: int = 40):
```
Missing return type.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+
+    calc_flags = recursive_dict_merge(calc_defaults, {})
+
+    try:
+        # Ensure the log directory exists
+        os.makedirs(logdir, exist_ok=True)
+
+        # Read reactant and product structures
+        reactant = read(xyz_r_p, index="0")
+        product = read(xyz_r_p, index="1")
+
+        # Optimize reactant and product structures using sella
+        for atom, name in zip([reactant, product], ["reactant", "product"]):
+            # atom.calc = calc()
+            atom.calc = NewtonNet(**calc_flags)
+            traj_file = os.path.join(logdir, f"{name}_opt.traj")
+            sella_wrapper(atom, traj_file=traj_file, sella_order=0)
+        # Save optimized reactant and product structures
+        r_p_path = os.path.join(logdir, "r_p.xyz")
+        write(r_p_path, [reactant.copy(), product.copy()])
+
+        # Generate intermediate images using geodesic interpolation
+        symbols, smoother_path = geodesic_interpolate_wrapper(
+            [reactant.copy(), product.copy()]
+        )
+        images = [Atoms(symbols=symbols, positions=conf) for conf in smoother_path]
+
+        # Calculate energies and forces for each intermediate image
+        for image in images:
+            # image.calc = calc()
```
Putting huge code in a `try`/`except` block is bad practice. You want to put only the minimal amount of code in the `try`/`except` block so you ensure you are catching what you expect to catch.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+
+    try:
+        # Ensure the log directory exists
+        os.makedirs(logdir, exist_ok=True)
```
This is not going to be possible because we cannot be writing out to custom directories.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+        return []
+
+
+def run_neb_method(
+    method: str,
+    optimizer: Optimizer | None = NEBOptimizer,
+    opt_method: str | None = "aseneb",
+    precon: str | None = None,
+    logdir: str | None = None,
+    xyz_r_p: str | None = None,
+    n_intermediate: int | None = 20,
+    k: float | None = 0.1,
+    max_steps: int | None = 1000,
+    fmax_cutoff: float | None = 1e-2,
+) -> None:
```
This should instead be added as `quacc.runners.ase.NEB` instead with a corresponding schema. That will also fix your problems with trying to write to custom directories.

Andrew-S-Rosen reviewed (commented):
Hi @kumaranu: I know this is still a WIP but I wanted to provide some initial comments. The biggest thing to note is that in quacc you cannot write things out to the current working directory or _any_ relative paths. Quacc does not `cd` into the working directory and so there is no guarantee that the current working directory or any relative path is going to be where the runtime/results directory is.

If you are looking to write out files before the calculation completes, then you need write it out based on the `tmpdir` that is made when `quacc.runners.pre.calc_setup` is called. For instance, in phonopy, you can see that certain files are written out (e.g. `phonopy.yaml`, `phonopy_auto_band_structure.yaml`) based on the `tmpdir` of the runtime directory:

https://github.com/Quantum-Accelerators/quacc/blob/04af31e4f8b64c386f35082570fe605879562c8d/src/quacc/runners/phonons.py#L55-L68

If you are writing to disk after the calculation completes and a schema is returned, you can query `schema["dir_name"]` to get the path where the results folder is. In any case, there is going to need to be a fairly significant amount of refactoring here to ensure that your I/O goes where you expect it to go. These functions may have been working fine for you because you are using FireWorks which `cd`'s into a custom directory for each run, but that is very unique to FireWorks. If you were to run these calculations without a workflow engine, for instance, you'll see that the files are not being written out to the unique results folder for that calculation.

Most of your issues in this PR boil down to the fact that you are trying to put too much into the recipes themselves when in reality what you are trying to do is make new `quacc.runners` functions that are called within your new recipes.

kumaranu commented on pyproject.toml:
```diff
@@ -49,7 +49,7 @@ defects = ["pymatgen-analysis-defects>=2023.8.22", "shakenbreak>=3.2.0"]
 mlp = ["matgl>=1.0.0", "chgnet>=0.3.3", "mace-torch>=0.3.3", "torch-dftd>=0.4.0"]
 mp = ["atomate2>=0.0.14"]
-newtonnet = ["newtonnet>=1.1"]
+newtonnet = ["newtonnet>=1.1", "geodesic-interpolate @ git+https://github.com/virtualzx-nad/geodesic-interpolate.git"]
```
Hi @Andrew-S-Rosen ,
I want to have tests that run for geodesic interpolate on GitHub. Is there a way you see to do that in case we are not able to put URLs?
Thanks.

Andrew-S-Rosen replied:
You will need to:
1. Update https://github.com/Quantum-Accelerators/quacc/blob/main/tests/requirements-newtonnet.txt to include the `geodesic-interpolate` direct link
2. Remove the direct link from pyproject.toml since we can't have direct links in there for PyPI

kumaranu commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -17,6 +26,11 @@
+try:
+    from ase.mep import neb
+except ImportError:
+    neb = None
```
I completely got rid of this import because I was not using it.
I think `ase` refactored their directory structure and moved the NEB functions to `ase.mep.neb`. So, I use:
`from ase.mep.neb import NEB`.
Thanks.

kumaranu commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -287,3 +301,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+
+"""
+@job
+@requires(NewtonNet, "NewtonNet must be installed. Refer to the quacc documentation.")
```
Changed this for all the `@requires` statements.

kumaranu commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+        qn.attach(traj.write, interval=traj_log_interval)
+    qn.run(fmax=fmax_cutoff, steps=max_steps)
+    if traj_file:
+        traj.close()
```
I removed `sella_wrapper` completely. It was something I was using when running things outside of `QuAcc`.

kumaranu commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+"""
+
+
+def sella_wrapper(
```
Haha! Yes, it should not have existed in the first place. I had it because I was running things independently and I moved the entire infrastructure to this branch in `QuAcc`. I will trying to replace things one by one making sure that the tests run after each modification.

kumaranu commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -293,3 +307,250 @@ def _get_hessian(atoms: Atoms) -> NDArray:
+"""
+@job
+@requires(NewtonNet, "NewtonNet must be installed. Refer to the quacc documentation.")
+def neb_job(
+    atoms: Atoms,
+    relax_endpoints: bool = True,
+    run_geodesic: bool = True,
+    run_single_ended: bool = True,
+    run_freq: bool = True,
+    neb_job_kwargs: dict[str, Any] | None = None,
+    relax_job_kwargs: dict[str, Any] | None = None,
+    freq_job_kwargs: dict[str, Any] | None = None,
+):
```
Skipped for now. I will get back to it.

kumaranu commented:
Hi @Andrew-S-Rosen
 
I have added a function in runners/ase.py with its test.

My test is running alright locally but it is complaining that the required packages are not installed such as "sella", "NewtonNet", "geodesic" etc.
https://github.com/Quantum-Accelerators/quacc/actions/runs/9449377214/job/26025492513?pr=2176

Please help resolve this issue. Thanks.

Tagging Sam here as I was discussing this issue with him recently.
@samblau

kumaranu commented:
> Hi @Andrew-S-Rosen
> 
> I have added a function in runners/ase.py with its test.
> 
> My test is running alright locally but it is complaining that the required packages are not installed such as "sella", "NewtonNet", "geodesic" etc. https://github.com/Quantum-Accelerators/quacc/actions/runs/9449377214/job/26025492513?pr=2176
> 
> Maybe the reason for this problem is that requirements-newtonnet.txt is not called for the packages that are present inside runners/ase.py. But I do not know which file will used to get that list of required packages.
> 
> Please help resolve this issue. Thanks.
> 
> Tagging Sam here as I was discussing this issue with him recently. @samblau

I added the following three lines to requirements.txt inside the tests directory to solve that problem:
newtonnet==1.1.1
geodesic-interpolate @ git+https://github.com/virtualzx-nad/geodesic-interpolate.git
sella==2.3.4

Andrew-S-Rosen commented on tests/core/recipes/orca_recipes/test_orca_recipes.py:
```diff
@@ -133,7 +133,7 @@ def test_ase_relax_job(tmp_path, monkeypatch):
         output["parameters"]["orcasimpleinput"]
         == "def2-tzvp engrad normalprint wb97x-d3bj xyzfile"
     )
-    assert output["parameters_opt"]["fmax"] == 0.1
```
Please don't change pre-existing tests. They should pass just fine.

Andrew-S-Rosen commented on pyproject.toml:
```diff
@@ -50,7 +50,7 @@ jobflow = ["jobflow[fireworks]>=0.1.14", "jobflow-remote>=0.1.0"]
 mlp = ["matgl>=1.1.2", "chgnet>=0.3.3", "mace-torch>=0.3.3", "torch-dftd>=0.4.0"]
 mp = ["atomate2>=0.0.14"]
 newtonnet = ["newtonnet>=1.1"]
-parsl = ["parsl[monitoring]>=2024.5.27; platform_system!='Windows'"]
```
Please don't change unrelated versions. They should work just fine.

kumaranu replied:
I don't know what was changed here. I see that this file is the same as the one in Quantum-Accelerator: main. Maybe you changed it back already.

kumaranu replied:
Did I change the parsl version from 2023.10.23 to 2.24.5.27?

Andrew-S-Rosen replied:
Yes, I took care of it. However, there are several other changes that still need to be reverted back, as shown in the file comparison view.

kumaranu commented:
> > Hi @Andrew-S-Rosen
> > I have added a function in runners/ase.py with its test.
> > My test is running alright locally but it is complaining that the required packages are not installed such as "sella", "NewtonNet", "geodesic" etc. https://github.com/Quantum-Accelerators/quacc/actions/runs/9449377214/job/26025492513?pr=2176
> > Maybe the reason for this problem is that requirements-newtonnet.txt is not called for the packages that are present inside runners/ase.py. But I do not know which file will used to get that list of required packages.
> > Please help resolve this issue. Thanks.
> > Tagging Sam here as I was discussing this issue with him recently. @samblau
> 
> I added the following three lines to requirements.txt inside the tests directory to solve that problem: newtonnet==1.1.1 geodesic-interpolate @ git+https://github.com/virtualzx-nad/geodesic-interpolate.git sella==2.3.4

However, the original problem is solved, I saw two new errors, tblite and mlp tests both of which I did not touch.

mlp test were just complaining that they could not find mace. I added the following lines in the requirements.txt to get rid of that error.
mace-torch==0.3.4
torch-dftd==0.4.0

However, the error in tblite seems much different as the tests ran fine but the numbers don't match. I have left that as it is:
https://github.com/Quantum-Accelerators/quacc/actions/runs/9456768276/job/26049315963?pr=2176

Andrew-S-Rosen commented:
@kumaranu: The `tests/requirements.txt` file cannot have the various packages you have added to it. It is meant to include a minimum set of packages corresponding to the `tests-core` CI run. If your tests require things like mace, they need to be running with the `tests-mlps` run. The reason the TBLite issues are happening is because TBLite is not compatible with torch (https://github.com/tblite/tblite/issues/110).

I do not quite understand what your goal is here with `tests/requirements.txt`. You have added NewtonNet tests, which should only run if NewtonNet is installed.

kumaranu commented:
> @kumaranu: The `tests/requirements.txt` file cannot have the various packages you have added to it. It is meant to include a minimum set of packages corresponding to the `tests-core` CI run. If your tests require things like mace, they need to be running with the `tests-mlps` run. The reason the TBLite issues are happening is because TBLite is not compatible with torch ([tblite/tblite#110](https://github.com/tblite/tblite/issues/110)).
> 
> I do not quite understand what your goal is here with `tests/requirements.txt`. You have added NewtonNet tests, which should only run if NewtonNet is installed.

I have removed NewtonNet and Sella from the test_ase.py requirements inside runners directory. It is using BFGS and EMT now.
@Andrew-S-Rosen @samblau

Andrew-S-Rosen commented:
@kumaranu: Thanks. I will review the PR once you undo all the unrelated changes you've made (there are still many) and remove commented out code blocks.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
Looks like these changes aren't needed anymore? They don't seem to be used anywhere.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -320,6 +326,153 @@ def run_vib(
     return vib
 
 
+def run_path_opt(
```
Let's rename this `run_neb`.

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -275,6 +275,40 @@ def summarize_vib_and_thermo(
     )
 
 
+def summarize_path_opt_run(dyn: Optimizer) -> OptSchema:
```
And then we can rename this `summarize_neb_run`

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -275,6 +275,40 @@ def summarize_vib_and_thermo(
+        "trajectory_results": [atoms.calc.results for atoms in trajectory],
+    }
+
+    # Create a dictionary of the inputs/outputs
+
+    # neb_dict = {
+    #     "images": [image.get_positions() for image in traj],
+    #     "fmax": neb.fmax,
+    #     "residuals": neb.residuals,
+    #     "k": neb.k,
+    #     "neb_method": neb.neb_method,
+    #     "nimages": neb.nimages,
+    #     "precon": neb.precon,
+    #     "optimizer": qn,
+    # }
+    # return neb_dict
```
FYI

kumaranu replied:
Got rid of that. Thanks.

Andrew-S-Rosen commented on tests/core/recipes/vasp_recipes/mocked/test_vasp_recipes.py:
Some more changes to revert later.

Andrew-S-Rosen commented on tests/core/schemas/test_ase.py:
Some more changes to revert later.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -320,6 +326,153 @@ def run_vib(
 
 
+def run_path_opt(
+    images,
```
Missing type hint.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -320,6 +326,153 @@ def run_vib(
+    copy_files: SourceDirectory | dict[SourceDirectory, Filenames] | None = None,
+) -> list[Atoms]:
+    """
+    Run NEB
+
+
+    Returns
+    -------
+    optimizer object
```
Docstring needs to be properly written.

Andrew-S-Rosen commented on src/quacc/__init__.py:
It seems that you may be making mistakes when trying to update your branch because your PR continually has code changes unrelated to your PR. For instance, in this file, there are code additions that are unrelated to the PR --- code that was since removed from `main`. Please make sure that when you are updating your branch, you are doing so correctly such that your PRs do not contain stray line changes.

kumaranu replied:
Hi @Andrew-S-Rosen 
This is a general problem. I am not sure how to check if a change in the main branch is to be accepted or rejected in the PR. When I click on update branch, I only check if the tests fail or pass right afterwards.
Also, I am not sure why is it not removing these lines when I have updated the code again.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    Perform a quasi-IRC job using the given atoms object. The initial IRC job by default
+    is run with `max_steps: 5`.
+
+    Parameters
+    ----------
+    atoms
+        The atoms object representing the system
+    direction
+        The direction of the IRC calculation
```
Your parameters don't match the function signature.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    relax_job_kwargs
+        Keyword arguments to use for the [quacc.recipes.newtonnet.core.relax_job][]
+
+    Returns
+    -------
+    A dictionary containing the neb optimization summary.
+            See the type-hint for the data structure.
+    """
```
This is not how a `Returns` is formatted. You need to have it like

```python
dict
    Description goes here
```

otherwise the docs do not build. This is called numpy docstring format.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    relax_summary_p = strip_decorator(relax_job)(product_atoms, **relax_job_kwargs)
+
+    images = _geodesic_interpolate_wrapper(
+        relax_summary_r["atoms"].copy(),
+        relax_summary_p["atoms"].copy(),
```
The copying should be done in the underlying function if necessary rather than beforehand.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    calc_kwargs: dict[str, Any] | None = None,
+    geodesic_interpolate_kwargs: dict[str, Any] | None = None,
+    neb_kwargs: dict[str, Any] | None = None,
+) -> dict:
```
Please use a more descriptive type hint than `dict`. In this case, you want to define a custom `TypedDict` class like

```python
NebSchema(TypedDict):
    relax_reactant: OptSchema
    relax_product: OptSchema
    ...
```

and then have this be the return type.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    neb_kwargs: dict[str, Any] | None = None,
+) -> dict:
+    """
+    Perform a quasi-IRC job using the given atoms object. The initial IRC job by default
+    is run with `max_steps: 5`.
```
This description does not seem accurate.

kumaranu replied:
Resolved.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    neb_kwargs: dict[str, Any] | None = None,
+) -> dict:
+    """
+    Perform a quasi-IRC job using the given atoms object. The initial IRC job by default
+    is run with `max_steps: 5`.
```
This description does not seem accurate.

kumaranu replied:
Resolved.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    calc_kwargs: dict[str, Any] | None = None,
+    geodesic_interpolate_kwargs: dict[str, Any] | None = None,
+    neb_kwargs: dict[str, Any] | None = None,
+) -> dict:
```
Please use a more detailed `TypedDict`-based type hint.

kumaranu replied:
I replaced it with a schema now. I hope that works.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+
+    Parameters
+    ----------
+    atoms
+        The atoms object representing the system
+    direction
+        The direction of the IRC calculation
```
Please update the docstring to match the function signature.

kumaranu replied:
Resolved.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    relax_job_kwargs
+        Keyword arguments to use for the [quacc.recipes.newtonnet.core.relax_job][]
+
+    Returns
+    -------
+    A dictionary containing the neb optimization summary.
+            See the type-hint for the data structure.
```
Please correct your return docstring.

kumaranu replied:
I guess this is resolved now.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    reactant_atoms.calc = NewtonNet(**calc_flags)
+    product_atoms.calc = NewtonNet(**calc_flags)
+
+    # Run IRC
+    relax_summary_r = strip_decorator(relax_job)(reactant_atoms, **relax_job_kwargs)
+    relax_summary_p = strip_decorator(relax_job)(product_atoms, **relax_job_kwargs)
```
I'm confused --- this doesn't seem like an IRC calculation.

kumaranu replied:
It was stupid typo I guess.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+
+    Parameters
+    ----------
+    reactant_atoms : Atoms
+        The atoms object representing the reactant state.
+    product_atoms : Atoms
+        The atoms object representing the product state.
+    relax_job_kwargs : dict[str, Any], optional
+        Keyword arguments to use for the relaxation job.
+    calc_kwargs : dict[str, Any], optional
+        Keyword arguments to configure the NewtonNet calculator.
+    geodesic_interpolate_kwargs : dict[str, Any], optional
+        Keyword arguments to configure the geodesic interpolation.
```
Please use the same docstring format as elsewhere in quacc, without the `:` and text afterwards.

kumaranu replied:
Did that.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    relax_job_kwargs: dict[str, Any] | None = None,
+    calc_kwargs: dict[str, Any] | None = None,
+    geodesic_interpolate_kwargs: dict[str, Any] | None = None,
+) -> dict:
```
Please use a more thorough type hint.

kumaranu replied:
Did that.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -264,6 +267,246 @@ def quasi_irc_job(
+    has_geodesic_interpolate,
+    "geodesic-interpolate must be installed. Refer to the quacc documentation.",
+)
+def geodesic_ts_job(
+    reactant_atoms: Atoms,
+    product_atoms: Atoms,
+    relax_job_kwargs: dict[str, Any] | None = None,
+    calc_kwargs: dict[str, Any] | None = None,
+    geodesic_interpolate_kwargs: dict[str, Any] | None = None,
+) -> dict:
```
There is a lot of the same code between your three newly added functions. You should consider refactoring them such that you do not introduce large blocks of duplicated code into the codebase.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -5,12 +5,16 @@
+from typing import TYPE_CHECKING, Any, Callable
 
 import numpy as np
+from ase import Atoms
+from ase.atoms import Atoms
```
Please only use one of these imports. I suggest the latter.

kumaranu replied:
Done.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+
+
+def run_neb(
+    images,
```
Missing type hint.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
             optimizer_kwargs["internal"] = True
+
+
+def run_neb(
+    images,
+    relax_cell: bool = False,
+    fmax: float = 0.01,
+    max_steps: int | None = 1000,
+    optimizer: NEBOptimizer | BFGS = NEBOptimizer,
+    optimizer_kwargs: OptimizerKwargs | None = None,
+    run_kwargs: dict[str, Any] | None = None,
+    neb_kwargs: dict[str, Any] | None = None,
+    copy_files: SourceDirectory | dict[SourceDirectory, Filenames] | None = None,
+) -> list[Atoms]:
```
At some point before merging, this function needs to be a method within the `Runner` class as you see for the other methods in this module. I can assist with that later, although you can of course try yourself if you wish.

kumaranu replied:
Did that.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+
+    Parameters
+    ----------
+    images : list of Atoms
+        List of images representing the initial path.
+    relax_cell : bool, optional
+        Whether to relax the unit cell shape and volume.
+    fmax : float, optional
+        Tolerance for the force convergence (in eV/A).
+    max_steps : int or None, optional
+        Maximum number of steps to take.
+    optimizer : Optimizer or BFGS, optional
+        Optimizer class to use.
+    optimizer_kwargs : dict or None, optional
+        Dictionary of kwargs for the optimizer.
+    run_kwargs : dict or None, optional
+        Dictionary of kwargs for the run() method of the optimizer.
+    neb_kwargs : dict or None, optional
+        Dictionary of kwargs for the NEB.
+    copy_files : str or dict or None, optional
```
Please update the docstring format to match the rest of quacc.

kumaranu replied:
Did that.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    relax_cell: bool = False,
+    fmax: float = 0.01,
+    max_steps: int | None = 1000,
+    optimizer: NEBOptimizer | BFGS = NEBOptimizer,
```
Is it true that only `NEBOptimizer` or `BFGS` can be used? That doesn't seem to be the case based on the docs. Looks like there are examples with `FIRE` for instance. The only thing I see in the docs is that `BFGSLineSearch` can't be used. Maybe you should just do `NEBOptimizer | Optimizer` and then raise a `ValueError` if the user passes an incompatible one (i.e. `BFGSLineSearch`)?.

kumaranu replied:
I tried GPMin and it gave RuntimeError. I have not tried others yet.

kumaranu replied:
I have made the requested change though.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    run_kwargs: dict[str, Any] | None = None,
+    neb_kwargs: dict[str, Any] | None = None,
+    copy_files: SourceDirectory | dict[SourceDirectory, Filenames] | None = None,
+) -> list[Atoms]:
```
This is not your return type.

kumaranu replied:
Corrected it.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    dyn.traj_atoms = read(traj_file, index=":")
+
+    # Perform cleanup operations
+    for ii, image in enumerate(images):
+        calc_cleanup(image, dir_lists[ii][0], dir_lists[ii][1])
```
You can repeat `i` here (instead of `ii`). That's fine.

kumaranu replied:
Did that.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+        raise ValueError(msg)
+
+    # Define the Trajectory object
+    traj_file = dir_lists[0][0] / "neb.traj"
```
Let's still call it the `opt.traj` for consistency.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+
+    # Set volume relaxation constraints, if relevant
+    for i in range(len(images)):
+        if relax_cell and images[i].pbc.any():
+            images[i] = FrechetCellFilter(images[i])
```
Probably no need to check for PBCs. If the user specifically requests `relax_cell=True`, that's on them.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+
+    Parameters
+    ----------
+    reactant_product_atoms : List[Atoms]
+        List of ASE Atoms objects containing initial and final geometries.
```
This is not matching your function signature.

kumaranu replied:
Fixed it earlier, probably. I could not find it now.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    return dyn
+
+
+def _geodesic_interpolate_wrapper(
```
This function needs a `@requires` decorator.

kumaranu replied:
Done.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+def _geodesic_interpolate_wrapper(
+    reactant: Atoms,
+    product: Atoms,
+    nimages: int = 20,
```
Please use `n_images` instead of `nimages`.

kumaranu replied:
Done.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    """
+    Interpolates between two geometries and optimizes the path with the geodesic method.
+
+    Parameters
+    ----------
+    reactant_product_atoms : List[Atoms]
+        List of ASE Atoms objects containing initial and final geometries.
```
As before, please update your docstrings to match the rest of quacc.

kumaranu replied:
Fixed.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    distance_cutoff : float, optional
+        Cut-off value for the distance between a pair of atoms to be included in the coordinate system. Default is 3.0.
+
+    Returns:
+    --------
```
Numpy-style docstrings do not have `Returns:` but rather just `Returns`. 

These and the other inconsistencies are why your docs test is failing in CI.

kumaranu replied:
Done.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+        chemical_symbols,
+        [reactant.positions, product.positions],
+        nimages,
+        tol=convergence_tolerance * 5,
```
Why do we multiply the user's tolerance by 5?

kumaranu replied:
This is from geodesic's own code.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+        threshold=distance_cutoff,
+        friction=geometry_friction,
+    )
+    if perform_sweep is None:
+        perform_sweep = len(chemical_symbols) > 35
```
Where does this 35 atom cutoff come from?

kumaranu replied:
From geodesic's own code.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    morse_scaling: float = 1.7,
+    geometry_friction: float = 1e-2,
+    distance_cutoff: float = 3.0,
+) -> tuple[list[str], list[list[float]]]:
```
It looks like your return is a `list[Atoms`], no?

kumaranu replied:
Yes! Fixed.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -362,3 +369,171 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+
+    Returns:
+    --------
+    Tuple[List[str], List[List[float]]]
+        A tuple containing the list of symbols and the smoothed path.
```
It looks like your return is a `list[Atoms]`.

kumaranu replied:
Fixed.

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -271,6 +271,40 @@ def summarize_vib_and_thermo(
     )
 
 
+def summarize_neb_run(dyn: Optimizer) -> OptSchema:
+    """
+    Summarizes the NEB run results into an OptSchema.
+
```
This function has several issues because it does not match how the other summarizers work. For instance, you'll see that the various summarizers return a `finalize_dict()` function where that function stores the results in a Maggma Store, among other things. Here, your results are never actually stored in your database. Please try to mimic what the other functions are doing.

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -271,6 +271,40 @@ def summarize_vib_and_thermo(
+        trajectory, and trajectory results.
+    """
+    # Get trajectory
+    trajectory = (
+        dyn.traj_atoms
+        if hasattr(dyn, "traj_atoms")
+        else read(dyn.trajectory.filename, index=":")
+    )
```
This logic has since been updated in `summarize_opt_run`. There is no `traj_atoms` attribute anymore. Please update to match what's being done in `summarize_opt_run`.

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -271,6 +271,40 @@ def summarize_vib_and_thermo(
     )
 
 
+def summarize_neb_run(dyn: Optimizer) -> OptSchema:
+    """
+    Summarizes the NEB run results into an OptSchema.
```
What does this function need to do differently than `summarize_opt_run`. Is there a reason why we can't use that function or simply call that function and make modifications on the output?

Note: You aren't returning an `OptSchema` here. You have made up your own dictionary.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+    return reactant, product
+
+
+@pytest.mark.parametrize(
+    (
+        "nimages",
+        "convergence_tolerance",
+        "max_iterations",
+        "max_micro_iterations",
+        "morse_scaling",
+        "geometry_friction",
+        "distance_cutoff",
+        "save_raw_path",
+        "expected_length",
+    ),
+    [
+        (20, 2e-3, 15, 20, 1.7, 1e-2, 3.0, None, 20),  # Default parameters
+        (10, 2e-3, 15, 20, 1.7, 1e-2, 3.0, None, 10),  # Different number of images
+        (
+            20,
+            1e-4,
+            10,
+            10,
+            1.5,
+            0.01,
+            2.5,
+            "raw_path.xyz",
+            20,
+        ),  # Different interpolation parameters and save_raw
+    ],
+)
```
Here and throughout the entire test suite changes, please do not use `pytest.mark.parameterize`. This is not the appropriate thing to do in your tests because it makes it extremely hard to tell at a quick glance what each parameter is doing. `@pytest.mark.parameterize` is for running a parameter sweep where you want to run the same test but with multiple values for a single parameter. It's not meant to run a single test. Please define the test function with no decorator and no arguments. Simply put the test values inside the function definition wherever they belong.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+        geometry_friction=geometry_friction,
+        distance_cutoff=distance_cutoff,
+    )
+    assert smoother_path[1].positions[0][0] == pytest.approx(1.36055556030, abs=1e-1)
```
This tolerance is _extremely_ large. Why are we using such a loose test here?

kumaranu replied:
I have made this tolerance so large because it is not deterministic. The value changes in every run. I am comparing a very approximate number here.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+
+
+def test_geodesic_interpolate_wrapper_large_system(setup_test_environment):
+    rng = np.random.default_rng()  # Create a random number generator instance
+    large_atoms = Atoms("H" * 40, positions=rng.random((40, 3)))
```
We do not want to introduce randomness into our test suite. Please define an `Atoms` object that is deterministic.

kumaranu replied:
This is because the results from Geodesic are not deterministic.
This is a serious limitation for testing.
@Andrew-S-Rosen @samblau 

There are two places in geodesic code where I found the use of random number generators. I can try and make it have a seed so that the results are reproducible. I have tried very quickly but not succeeded so far.

Andrew-S-Rosen replied:
Even if geodesic-interpolate is stochastic, why would we need a random `Atoms` object to provide?

kumaranu replied:
I have replaced it with C60. Thanks for the suggestion.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+        "last_images_positions",
+    ),
+    [
+        # ("aseneb", NEBOptimizer, None, 10, 0.1, 3, 1e-3, None),
+        # ("aseneb", SciPyFminBFGS, None, 1000, 0.1, 3, 1e-3, "some_logdir",
+        #   0.78503956131, -24.9895786292, -0.0017252843, 0.78017739462, 9, -19.946616164,
+        #   -0.19927549, 0.51475535802),
```
Please remove all commented out code.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+        ),  # Different interpolation parameters and save_raw
+    ],
+)
+def test_geodesic_interpolate_wrapper(
+    setup_test_environment,
+    nimages,
+    convergence_tolerance,
+    max_iterations,
+    max_micro_iterations,
+    morse_scaling,
+    geometry_friction,
+    distance_cutoff,
+    save_raw_path,
+    expected_length,
+):
```
This test needs to be skipped if `geodesic-interpolate` is not installed. Please use a `@pytest.mark.skipif` decorator.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+    assert smoother_path[1].positions[0][0] == pytest.approx(1.36055556030, abs=1e-1)
+
+
+def test_geodesic_interpolate_wrapper_large_system(setup_test_environment):
```
This test needs to be skipped if `geodesic-interpolate` is not installed. Please use a `@pytest.mark.skipif` decorator.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+        )
+    ],
+)
+def test_run_neb(
```
Your test of `run_neb` should not rely on `geodesic-interpolate`. If you want to test the NEB method, generate the images using ASE's methods. This will ensure we are only testing one thing at a time (hence the name a "unit test").

kumaranu replied:
The NEB function needs list of ASE images. I am avoiding it because it will just become a long list of coordinates in this file. Not sure how to do it cleanly.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+):
+    reactant, product = setup_test_environment
+
+    optimized_r = strip_decorator(relax_job)(reactant)["atoms"]
+    optimized_p = strip_decorator(relax_job)(product)["atoms"]
```
Please don't run unnecessary calculations in your tests. We want tests to evaluate only one thing. Simply provide a pre-made set of images to test the `run_neb` method.

Andrew-S-Rosen replied:
Yes, you can either add `.xyz` geometries and read them in or define an `Atoms` object that has the positions, atom identities, etc. of the structure you want to be testing here.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -47,6 +52,196 @@ def teardown_function():
+    dyn = run_neb(images, optimizer=NEBOptimizer, neb_kwargs=neb_kwargs)
+    neb_summary = summarize_neb_run(dyn)
+
+    assert neb_summary["trajectory_results"][1]["energy"] == pytest.approx(
+        1.098, abs=0.01
+    )
```
A more thorough test of the output schema is required.

kumaranu replied:
Same problem as the geodesic here because I am using geodesic as the precursor to this step.

Andrew-S-Rosen replied:
There are too many things being tested here at once is the major problem. For instance, `runners/test_ase.py` should be testing the runner: `run_neb`. It should be asserting that `run_neb()` functions appropriately and nothing else. There should be no precursor to that step and nothing afterwards (i.e. no summarizing).

You'd then make a separate test for summarizing (e.g. `tests.cores.schemas`) and so on and so forth.

Andrew-S-Rosen commented on tests/core/schemas/test_ase.py:
Please do not make unrelated changes in your PR.

Andrew-S-Rosen reviewed (commented):
Hi @kumaranu. I have provided an initial review below. There are many areas for cleanup, but overall the logic seems mostly reasonable. If you can address these changes, that would be ideal. Please note that you should also be checking GitHub Actions CI to see what tests fail. You'll see that your docs and pre-commit CI runners failed. Clicking the "details" will show you why.

Once you make these changes, I will review it once more and can assist with remaining issues. Thank you. Please also "resolve" your comments when you are done addressing them.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -48,6 +56,182 @@ def teardown_function():
+    reason="geodesic_interpolate function is not available",
+)
+def test_geodesic_interpolate_wrapper_large_system(setup_test_environment):
+    rng = np.random.default_rng()  # Create a random number generator instance
```
This is highly stochastic. If you want it to be deterministic, provide a `seed`, e.g. `seed=42`.

kumaranu commented:
Hi @Andrew-S-Rosen 

I am not sure why one of my tests is failing on Windows all of a sudden. Please let me know if you see anything that is obvious to you. Thanks.

Andrew-S-Rosen commented:
It looks like a tolerance issue and should be not a problem. You can change the tolerance slightly (just enough for the tests to pass). I am not sure the source either.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -266,6 +282,269 @@ def quasi_irc_job(
+    has_geodesic_interpolate,
+    "geodesic-interpolate must be installed. Refer to the quacc documentation.",
+)
+def geodesic_ts_job(
+    reactant_atoms: Atoms,
+    product_atoms: Atoms,
+    relax_job_kwargs: dict[str, Any] | None = None,
+    calc_kwargs: dict[str, Any] | None = None,
+    geodesic_interpolate_kwargs: dict[str, Any] | None = None,
+) -> NebTsSchema:
+    """
+    Perform a quasi-IRC job using the given reactant and product atoms objects.
```
The docstring does not match the name. Is it TS or quasi-IRC.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -42,6 +45,19 @@ class QuasiIRCSchema(OptSchema):
         irc_job: IRCSchema
         freq_job: FreqSchema | None
 
+    class NebSchema(TypedDict):
+        relax_reactant: OptSchema
+        relax_product: OptSchema
+        geodesic_results: list[Atoms]
+        neb_results: dict
+
+    class NebTsSchema(TypedDict):
+        relax_reactant: OptSchema
+        relax_product: OptSchema
+        geodesic_results: list[Atoms]
+        neb_results: dict
+        ts_results: TSSchema
```
The `neb_results: dict` should use the type hint for the return from `summarize_neb_run` (if we keep that function) rather than `dict`.

kumaranu replied:
Not sure how to do this one. I see that quacc.types has the base classes to define the schemas but I don't have those for summarize_neb_run.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -266,6 +282,269 @@ def quasi_irc_job(
     return relax_summary
 
 
+@job
+@requires(
+    has_newtonnet, "NewtonNet must be installed. Refer to the quacc documentation."
+)
+@requires(
+    has_geodesic_interpolate,
+    "geodesic-interpolate must be installed. Refer to the quacc documentation.",
+)
+def neb_job(
+    reactant_atoms: Atoms,
+    product_atoms: Atoms,
+    relax_job_kwargs: dict[str, Any] | None = None,
+    calc_kwargs: dict[str, Any] | None = None,
+    geodesic_interpolate_kwargs: dict[str, Any] | None = None,
+    neb_kwargs: dict[str, Any] | None = None,
+) -> NebSchema:
```
This comment applies to both recipes:

Like we did with sella and optimizations, we are going to want all the recipes here to be run without geodesic-interpolate if desired. geodesic-interpolate is merely one of many interpolation methods, and these functions do not strictly rely on geodesic-interpolate for functionality --- they are only relying on it to generate a set of images. Many methods can do that. The main concern is that one day, geodesic-interpolate _will_ stop working and break (because it's not actively maintained). We want to make sure the recipes can still be used if/when that happens, or even in cases where better methods come along.

To do this, I would suggest:
- Removing the `@requires` decorator for geodesic-interpolate
- Having a keyword argument that's like `interpolation_method: Literal["linear", "geodesic"] = "linear"`. If set to `"geodesic"`, then call your geodesic-interpolate wrapper. If set to `"linear"`, use ASE's default interpolate method. One can easily see how this could be expanded later to include idpp or others.
- Change `geodesic_interpolate_kwargs` to `interpolation_kwargs`, which would be the kwargs passed to whatever interpolation method is used (i.e. geodesic's or ASE's).

With this approach, it will be more flexible. Imagine one day you make a new image interpolation scheme `anup_image_gen`. We could then add it to `interpolation_method: Literal["linear", "geodesic", "anup"]` without having to make a new recipe.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -266,6 +282,269 @@ def quasi_irc_job(
+        The Atoms object representing the reactant structure.
+    product_atoms
+        The Atoms object representing the product structure.
+    relax_job_kwargs
+        Keyword arguments to use relax_job.
+    calc_kwargs
+        Custom kwargs for the NewtonNet calculator.
+    geodesic_interpolate_kwargs
+        Keyword arguments for the geodesic function.
+    neb_kwargs
+        Keyword arguments for the NEB calculation.
```
This comment applies to both functions:

As done in other recipes throughout quacc, please clearly cite which functions or classes you are referring to. For instance, "Keyword arguments to use [quacc.recipes.newtonnet.core.relax_job][]` is much clearer than just "relax_job." It will also automatically hyperlink in the built docs. Do this anytime you have a "kwargs" as an argument.

kumaranu replied:
I changed a few of them. Some of them like the geodesic interpolate kwargs did not work and gave errors in docs so I got rid of them.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -266,6 +282,269 @@ def quasi_irc_job(
     return relax_summary
 
 
+@job
+@requires(
+    has_newtonnet, "NewtonNet must be installed. Refer to the quacc documentation."
+)
+@requires(
+    has_geodesic_interpolate,
+    "geodesic-interpolate must be installed. Refer to the quacc documentation.",
+)
+def neb_job(
+    reactant_atoms: Atoms,
+    product_atoms: Atoms,
+    relax_job_kwargs: dict[str, Any] | None = None,
+    calc_kwargs: dict[str, Any] | None = None,
+    geodesic_interpolate_kwargs: dict[str, Any] | None = None,
+    neb_kwargs: dict[str, Any] | None = None,
+) -> NebSchema:
```
I haven't commented on this previously but really should now:

In quacc, the `@job` is a discrete unit of work. A `@flow` is a workflow, concatenating multiple distinct `@job`s. Here (and in your other recipes, for what it's worth), you are actually doing a workflow. You have two relaxations (two "jobs") before running an NEB calculation. 

Normally, I would say this is not permitted. If we were running a Q-Chem calculation for instance, this would be highly undesirable because you would want the reactant and product relaxation jobs to be done concurrently, you'd want those results to be stored in your database, and if one fails you'd still like to store the results from the other. You can't do that if everything is packed in one `@job`.

The challenge here is that because NewtonNet and MLPs are so fast, it is tempting to pack everything into one `@job`. For users of workflow engines (like FireWorks) where each `@job` is a Slurm job, this means you'd be queuing more than you need. But not all workflow engines work that way (and even FireWorks has the option to not re-queue each job).

Jobflow does not support the `@flow` pattern, so I'm not going to inconvenience you by requiring that you implement it here. But I wanted to bring up this difference: what you are actually doing here is making a flow that combines steps from multiple jobs.

kumaranu replied:
I think the recipes like neb_ts_job and geodesic_ts_job will most likely go away. I have left them working right now but will probably remove them after a discussion together. I have put the workflows on a separate repository so things like the flow can be done over there. I have however not use jobflow over there so the parallel nature of the execution of the codes and also the fireworks related functionalities don't exist in those workflows as well.

Andrew-S-Rosen reviewed (commented):
Hi @kumaranu and @samblau: Please see my comments below. Thank you again for your contribution, and I hope that this helps.

Andrew-S-Rosen commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -266,6 +282,269 @@ def quasi_irc_job(
+    reactant_atoms: Atoms,
+    product_atoms: Atoms,
+    relax_job_kwargs: dict[str, Any] | None = None,
+    calc_kwargs: dict[str, Any] | None = None,
```
This applies to both recipes:

I would suggest making this `**calc_kwargs` as done in the other recipes you made.

Andrew-S-Rosen commented on tests/core/recipes/newtonnet_recipes/test_newtonnet_recipes.py:
Are you testing specific functions locally? That would be much faster than relying on GitHub Actions. I saw this commit, which looked like you were trying to debug via the git history.

kumaranu replied:
I do test the functions locally but unfortunately the random seed messes up things on GitHub actions. It is because it is not reliable across different python versions.

samblau commented on tests/core/runners/test_ase.py:
```diff
@@ -3,21 +3,36 @@
 import pytest
+
+
+@pytest.fixture(scope="session", autouse=True)
+def set_seed():
+    np.random.seed(42)  # noqa: NPY002
```
So the issue is `scope="session"`? It's not entirely clear to me how to restrict this to module scope as you suggest.

kumaranu commented on tests/core/runners/test_ase.py:
```diff
@@ -48,6 +63,180 @@ def teardown_function():
+
+    reactant, product = setup_test_environment
+
+    optimized_r = strip_decorator(relax_job)(reactant)["atoms"]
+    optimized_p = strip_decorator(relax_job)(product)["atoms"]
```
I have removed the strip_decorator calls from the test_ase file but I am still confused a bit. Should I use strip_decorator is I am calling relax_job in a script that does not use a workflow engine. I mean if I just want to run the things locally but not in a jobflow like format.

kumaranu commented on tests/core/runners/test_ase.py:
```diff
@@ -48,6 +63,180 @@ def teardown_function():
+    n_intermediate = 10
+    r_positions = -0.849607247104427
+    p_energy = 1.0824716056541726
+    first_image_forces = -0.0052292931195385695
```
Removed the other calls now. Directly reading the geodesic path from a file now.

kumaranu commented on src/quacc/schemas/ase.py:
```diff
@@ -121,6 +121,55 @@ def summarize_run(
+    store: Store | None = _DEFAULT_SETTING,
+) -> RunSchema:
+    """
+    Summarize the NEB run results and store them in a database-friendly format.
```
Removed this function and replaced the relevant call with atoms_to_metadata.

kumaranu commented on src/quacc/schemas/ase.py:
```diff
@@ -348,6 +397,72 @@ def summarize_vib_and_thermo(
+        traj_atoms.calc = None
+
+    initial_atoms = trajectory[0]
+    get_final_atoms_from_dynamics(dyn)
```
Got rid of this get_final_atoms_from_dynamics call.

kumaranu commented on src/quacc/schemas/ase.py:
```diff
@@ -348,6 +397,72 @@ def summarize_vib_and_thermo(
+    # Base task doc
+    base_task_doc = summarize_run_neb2(
+        initial_atoms, charge_and_multiplicity=charge_and_multiplicity, store=None
+    )
```
As we talked during a zoom meeting, there are some problems with calling summarize_opt_run as it is right now. I will address other comments first and maybe get back to this part later on.

kumaranu replied:
summarize_run_neb2 has been removed though.

kumaranu commented on tests/core/runners/test_ase.py:
```diff
@@ -48,6 +63,180 @@ def teardown_function():
+)
+def test_run_neb2(setup_test_environment, tmp_path):
+    """
+    Test the NEB calculation with geodesic interpolation.
```
Removed the unnecessary part and now reading in the geodesic path directly.

kumaranu commented on tests/core/runners/test_ase.py:
Removed the geodesic skip statements.

kumaranu commented on tests/requirements.txt:
```diff
@@ -11,3 +11,4 @@ pydantic-settings==2.3.4
 pymatgen==2024.6.10
 ruamel.yaml==0.18.6
 typer==0.12.3
+geodesic-interpolate @ git+https://github.com/virtualzx-nad/geodesic-interpolate.git
```
Did that!

kumaranu commented on src/quacc/runners/ase.py:
```diff
@@ -453,3 +459,188 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    run_kwargs: dict[str, Any] | None = None,
+    neb_kwargs: dict[str, Any] | None = None,
+    copy_files: SourceDirectory | dict[SourceDirectory, Filenames] | None = None,
+) -> Optimizer:
```
I updated the returned type-hint to Dynamics and also updated the docstring for that as well.

kumaranu commented on src/quacc/runners/ase.py:
```diff
@@ -453,3 +459,188 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+        if relax_cell and images[i].pbc.any():
+            images[i] = FrechetCellFilter(images[i])
+
+    # Run optimization
+    dyn = optimizer(neb, **optimizer_kwargs)
+    dyn.attach(traj.write)
+    dyn.run(fmax, max_steps)
+    traj.close()
```
I am skipping this one for now. I know that this is tricky depending upon the optimizer. Maybe I will get back to this one later.

kumaranu commented on src/quacc/runners/ase.py:
```diff
@@ -453,3 +459,188 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    dyn.run(fmax, max_steps)
+    traj.close()
+
+    traj.filename = traj_file
+    dyn.trajectory = traj
```
I moved it to after the clean up. Thanks for clarifying this part. This was before I realized that the tmp files are renamed to quacc-* files instead of getting deleted.

kumaranu commented on src/quacc/runners/ase.py:
```diff
@@ -453,3 +459,188 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    dyn.trajectory = traj
+
+    # Perform cleanup operations skipping the first images's directory
+    # because that is where the trajectory is stored. It will get deleted
+    # eventually.
```
Got rid of that condition now.

kumaranu commented on src/quacc/runners/ase.py:
```diff
@@ -453,3 +459,188 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+        chemical_symbols,
+        [reactant.positions, product.positions],
+        n_images,
+        tol=convergence_tolerance * 5,
```
I think the code was trying to use the same number in the relative sense that the smooth or sweep functions to have a tighter convergence criteria as compared to the redistribute function.
Nevertheless, I agree with you that this multiplication by 5 is quite weird.
So, I got rid of that and made separate arguments for the two types of tolerances now.

kumaranu commented on src/quacc/runners/ase.py:
```diff
@@ -453,3 +459,188 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    run_kwargs: dict[str, Any] | None = None,
+    neb_kwargs: dict[str, Any] | None = None,
+    copy_files: SourceDirectory | dict[SourceDirectory, Filenames] | None = None,
+) -> Optimizer:
```
OK.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -415,3 +421,194 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    return dyn
+
+
+@requires(
+    has_geodesic_interpolate,
+    "geodesic-interpolate must be installed. Refer to the quacc documentation.",
+)
+def _geodesic_interpolate_wrapper(
```
To me, it looks like this function belongs in `quacc.atoms` and not here. Maybe you can store it in a new file `quacc.atoms.ts`?

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -349,6 +350,113 @@ def summarize_vib_and_thermo(
+    if not trajectory:
+        trajectory = read(dyn.trajectory.filename, index=":")
+
+    n_images = additional_fields["interpolate_flags"]["n_images"]
+
+    n = additional_fields.get("n_iter_return", -1)
```
This is going to be a problem because `additional_fields` is not meant to be accessed this way. The `additional_fields` argument is meant to be purely optional --- we can't rely on it being specified.

Perhaps you can add `n_images` as a required positional argument if you need to access this? And `n_iter_return` as an optional keyword argument?

kumaranu replied:
I think I resolved it.

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -349,6 +350,113 @@ def summarize_vib_and_thermo(
+    )
+    ts_atoms = trajectory[ts_index]
+
+    for traj_atoms in trajectory:
+        traj_atoms.calc = None
```
Not sure this is needed?

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -349,6 +350,113 @@ def summarize_vib_and_thermo(
+    additional_fields = additional_fields or {}
+
+    # Get trajectory
+    if not trajectory:
+        trajectory = read(dyn.trajectory.filename, index=":")
```
To prevent overwriting of variables, can you make this

```python
if trajectory:
    atoms_trajectory = trajectory
else:
    atoms_trajectory =  read(dyn.trajectory.filename, index=":")  # type: ignore[union-attr]
```

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -349,6 +350,113 @@ def summarize_vib_and_thermo(
+        )
+    trajectory_results = [atoms.calc.results for atoms in trajectory]
+    ts_index = (
+        np.argmax([i["energy"] for i in trajectory_results[-(n_images - 1) : -1]]) + 1
```
Can you replace `i` with `results` (or something similar) to make this more descriptive?

samblau replied:
Probably `result` rather than `results`, since this would be one entry in `traj_results`

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -349,6 +350,113 @@ def summarize_vib_and_thermo(
+    """
+    Extract every nth iteration from the NEB trajectory.
+
+    Parameters:
+    - neb_trajectory: List of configurations (length: n_iter * n_images)
+    - n_iter: Total number of iterations
+    - n_images: Number of images per iteration
+    - n: Interval to get every nth iteration
+
+    Returns:
+    - List of configurations from every nth iteration
```
Can you use numpy docstyle docstrings like the other functions?

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -349,6 +350,113 @@ def summarize_vib_and_thermo(
+    - neb_trajectory: List of configurations (length: n_iter * n_images)
+    - n_iter: Total number of iterations
+    - n_images: Number of images per iteration
+    - n: Interval to get every nth iteration
```
Could you rename `n` to something more descriptive? Maybe `interval`?

Andrew-S-Rosen reviewed (commented):
Thanks, @kumaranu! I have left several comments here. While there are admittedly many comments, none of them are particularly challenging to address.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -415,3 +421,194 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    for i, image in enumerate(images):
+        calc_cleanup(image, dir_lists[i][0], dir_lists[i][1])
+
+    traj.filename = zpath(dir_lists[0][1] / traj_filename)
```
I think you need to do `zpath(str(dir_lists[0][1] / traj_filename))` to fix the CI errors you're getting. It's from an upstream change to monty.

kumaranu replied:
Done. Thanks.

samblau commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -259,6 +276,397 @@ def quasi_irc_job(
+    traj_results = neb_results["neb_results"]["trajectory_results"]
+    n_images = len(neb_results["geodesic_results"])
+
+    ts_index = np.argmax([i["energy"] for i in traj_results[-(n_images - 1) : -1]]) + 1
```
Probably `result` rather than `results`, since this would be one entry in `traj_results`

Andrew-S-Rosen replied:
Good point.

samblau commented on src/quacc/schemas/ase.py:
```diff
@@ -349,6 +350,113 @@ def summarize_vib_and_thermo(
+    parameters_opt.pop("restart", None)
+
+    opt_fields = {
+        "highest_e_atoms": ts_atoms,
```
Even if the NEB converges, the highest energy image is not necessarily a TS. Technically you always need to do a TSopt on the highest energy image. Perhaps just a matter of semantics...

Andrew-S-Rosen replied:
This is true. It would only be a TS with the climbing image variant. 

We can keep it as-is then. The key "highest_e_atoms" is a bit awkward, but I can't really think of a better name to be honest.

kumaranu commented on .github/check_env.yml:
Done. Thanks for the suggestion about "Files changed" tab. I will have a look at that.

kumaranu commented on src/quacc/recipes/newtonnet/core.py:
```diff
@@ -217,11 +221,13 @@ def _add_stdev_and_hess(summary: dict[str, Any]) -> dict[str, Any]:
         Hessian values.
     """
     settings = get_settings()
+    calc_defaults = {
+        "model_path": settings.NEWTONNET_MODEL_PATH,
+        "settings_path": settings.NEWTONNET_CONFIG_PATH,
+    }
+    calc_flags = recursive_dict_merge(calc_defaults, calc_kwargs)
```
Addressed it.

kumaranu commented:
Hi @Andrew-S-Rosen 

I am struggling with the path of traj file for neb. The error is only present on GitHub but not when I run the tests locally.
Please let me know if you see any problems over here.
https://github.com/Quantum-Accelerators/quacc/actions/runs/10085853797/job/27887414542?pr=2176
I have not gone through all of your comments so I apologize for asking if you already mentioned it somewhere else. Thanks.

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -126,9 +126,9 @@ def test_geodesic_interpolate_wrapper_large_system(setup_test_environment):
+    geodesic_path = str(os.path.join(
         os.path.dirname(os.path.abspath(__file__)), "geodesic_path.xyz"
-    )
+    ))
```
This change is not needed.

kumaranu replied:
I have undone this change. Thanks.

Andrew-S-Rosen commented:
> Hi @Andrew-S-Rosen
> 
> I am struggling with the path of traj file for neb. The error is only present on GitHub but not when I run the tests locally. Please let me know if you see any problems over here. https://github.com/Quantum-Accelerators/quacc/actions/runs/10085853797/job/27887414542?pr=2176 I have not gone through all of your comments so I apologize for asking if you already mentioned it somewhere else. Thanks.

Please see this comment: https://github.com/Quantum-Accelerators/quacc/pull/2176#discussion_r1690015953.

kumaranu commented:
> > Hi @Andrew-S-Rosen
> > I am struggling with the path of traj file for neb. The error is only present on GitHub but not when I run the tests locally. Please let me know if you see any problems over here. https://github.com/Quantum-Accelerators/quacc/actions/runs/10085853797/job/27887414542?pr=2176 I have not gone through all of your comments so I apologize for asking if you already mentioned it somewhere else. Thanks.
> 
> @kumaranu: Please see this comment: [#2176 (comment)](https://github.com/Quantum-Accelerators/quacc/pull/2176#discussion_r1690015953).

Done this. All checks are passing now. Thanks.

kumaranu commented on src/quacc/recipes/newtonnet/ts.py:
```diff
@@ -259,6 +276,397 @@ def quasi_irc_job(
+    return {
+        "relax_reactant": relax_summary_r,
+        "relax_product": relax_summary_p,
+        "geodesic_results": images,
```
I agree with you. Now, that we have other methods included to provide initial guesses for the path to NEB, this should be called geodesic_results. I have changed it to initial_images all through.

kumaranu commented on src/quacc/runners/ase.py:
```diff
@@ -415,3 +421,194 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    Dynamics
+        The ASE Dynamics object following an optimization.
+    """
+    if optimizer.__name__ == "BFGSLineSearch":
```
I had not imported BFGSLineSearch ealier. Now, I imported it at the top and changed the if condition to how you suggested.

kumaranu commented:
Hi @Andrew-S-Rosen,

Thanks a lot for providing the comments above. I have finished going through them all.

As you mention two things where you might need to pitch in:
1. run_neb method in the Runner class
2. summarize_opt_run with sumarize_neb_run.

I have left those two unresolved.

Tagging @samblau here just for the update.

Andrew-S-Rosen commented:
Thanks, @kumaranu! I recognize that this has likely taken longer than you anticipated and involved more iterations than desired. That said, I feel that the end result here is much more robust and will ensure that your work can continue to be well-maintained into the future.

I am starting my new position Aug. 1st so can't wrap this up immediately, but I will keep it on my to-do list for the near future (middle of August most likely). As you suggested, the two things that need to happen before the final merge are: 1) Ensuring that the `run_neb` function becomes a method of the `Runner` class rather than a standalone function; 2) Ensuring that the `summarize_neb_run` function become a method of the `Summarize` class rather than a standalone function. Both scenarios here are meant to minimize code duplication so that changes to one part of the code are properly reflected in your NEB runner and schema. I will ping you when that's done so you can look it over before we wrap this up.

samblau commented:
hey @kumaranu , any progress on this? Thanks!

Andrew-S-Rosen commented on tests/core/runners/test_ase.py:
```diff
@@ -306,7 +306,7 @@ def test_run_neb(monkeypatch, tmp_path):
 
     assert traj[-1].calc.results is not None
-    assert not os.path.exists(tmp_path / "opt.log")
+    assert os.path.exists(tmp_path / "opt.log")
```
@kumaranu: I intentionally added this test. There should be no I/O happening in the runtime directory. This is not the directory where the calculations are stored.

kumaranu replied:
I see.
@Andrew-S-Rosen ,
I am not sure if my constraining the optimization with `max_steps=10` caused this test to fail or if it was already failing.

Andrew-S-Rosen replied:
I added the test to highlight the problem and so that we don't merge the PR before it's fixed. It should not have anything to do with your `max_steps` addition.

Andrew-S-Rosen replied:
Please revert it and patch before the merge, thanks.

kumaranu replied:
Reverted that.

Andrew-S-Rosen commented on src/quacc/runners/ase.py:
```diff
@@ -402,3 +405,93 @@ def _set_sella_kwargs(self, optimizer_kwargs: dict[str, Any]) -> None:
+    neb_kwargs = neb_kwargs or {}
+    traj_filename = "opt.traj"
+    optimizer_kwargs = recursive_dict_merge(
+        {"logfile": "opt.log", "restart": "opt.json"}, optimizer_kwargs
```
@kumaranu here is the issue. You are writing these log files out to the current working directory, which should be kept clean. In the run_opt method, you'll see this file is written out to the temporary directory where the calculation results are stored. You don't have that kind of setup here with your NEB run. I wonder if it makes sense to have one "parent" folder for the NEB run and then each image can be a subfolder within that.

kumaranu replied:
Hi @Andrew-S-Rosen ,
I have added a directory structure to the neb run here. I hope this works. Thanks.

kumaranu commented:
> hey @kumaranu , any progress on this? Thanks!

Hi @samblau ,
Things have been quite busy lately. Sorry, I couldn't get back to it earlier.

I reduced the timing of the tests by adding max_iter to them and also got rid of writing to the current directory problem. Let's see if it works out OK or not. Thanks!

Andrew-S-Rosen commented on src/quacc/schemas/ase.py:
```diff
@@ -516,3 +600,39 @@ def vib_and_thermo(
+        List of configurations from every nth iteration.
+    """
+    result = []
+    if interval > n_iter:
+        pass
```
@kumaranu: FYI that this code block doesn't actually do anything. I'm going to remove it, but if you meant for it to do something else just let me know.

Andrew-S-Rosen commented:
Alright friends, @kumaranu and @samblau. It is time for the merge! Is it perfect? No. Is it good enough? We'll find out! 

I did a _lot_ of refactoring here, but things seem to be okay. Let's move.
